### PR TITLE
drivers/tca64xx: Add support for PCAL6416

### PR DESCRIPTION
**Pull request by joukkone**
New part support added to driver,
pullup/pulldown configuration support

### CODE REVIEW

File: drivers/ioexpander/tca64xx.c [hunk 1]
```diff
@@ -439,6 +482,55 @@ static int tca64_direction(FAR struct ioexpander_dev_s *dev, uint8_t pin,
               regaddr, ret);
     }
 
+  if ((direction == IOEXPANDER_DIRECTION_IN_PULLDOWN) ||
+     (direction == IOEXPANDER_DIRECTION_IN_PULLUP))
+    {
+      regaddr = tca64_pdenable_reg(priv, pin);
```
Comment (jlaitine) [inline drivers/ioexpander/tca64xx.c:488]: There is no error handling now if PU/PD setting is attempted for other parts than the one which has the register. You should propably add "define INVALID_REG  0xff" or such, and make tca64_pdenable_reg return that if part->tp_pu_select is 0.

Then you can check the regaddr here and bail out if the part doesn't support PU/PD....

Comment (joukkone) [reply drivers/ioexpander/tca64xx.c:504]: It should be fixed now.

File: drivers/ioexpander/tca64xx.c [hunk 2]
```diff
@@ -254,6 +273,50 @@ static uint8_t tca64_config_reg(FAR struct tca64_dev_s *priv, uint8_t pin)
...
+ *  Return the address of the pu/pd enable register for the specified pin.
+ *
+ ****************************************************************************/
+
+static uint8_t tca64_pdenable_reg(FAR struct tca64_dev_s *priv, uint8_t pin)
+{
+  FAR const struct tca64_part_s *part = tca64_getpart(priv);
+  uint8_t reg = part->tp_puenable;
+
+  if (reg == TC64XX_INVALID_REG)
```
Comment (jlaitine) [inline drivers/ioexpander/tca64xx.c:289]: If you do other way around;

if (reg === 0) return TC64XX_INVALID_REG;

This makes more sense. Now this  returns 0, which is a valid register address.... even though never pdenable register.

File: drivers/ioexpander/tca64xx.c [hunk 3]
```diff
@@ -439,6 +498,81 @@ static int tca64_direction(FAR struct ioexpander_dev_s *dev, uint8_t pin,
               regaddr, ret);
     }
 
+  if ((direction == IOEXPANDER_DIRECTION_IN_PULLDOWN) ||
+     (direction == IOEXPANDER_DIRECTION_IN_PULLUP))
+    {
+      regaddr = tca64_pdenable_reg(priv, pin);
+      if (regaddr == 0)
```
Comment (jlaitine) [inline drivers/ioexpander/tca64xx.c:505]: And here if (regaddr ==  TC64XX_INVALID_REG) ....

File: drivers/ioexpander/tca64xx.c [hunk 4]
```diff
@@ -132,6 +135,8 @@ static const struct tca64_part_s g_tca64_parts[TCA64_NPARTS] =
     TCA6408_OUTPUT_REG,
     TCA6408_POLARITY_REG,
     TCA6408_CONFIG_REG,
+    TC64XX_INVALID_REG,
```
Comment (jlaitine) [inline drivers/ioexpander/tca64xx.c:138]: These are not needed if done  as suggested below

**Comment by jlaitine** [comment]
What you did works; it is just a bit confusing because it returns regaddr, which is valid (0), and you just know that 0 cannot be right one. Rather return the invalid address all the way to the place where it is queried.

File: drivers/ioexpander/tca64xx.c (hunk 2 above)
Comment (joukkone) [reply drivers/ioexpander/tca64xx.c]: done

**Review by jlaitine — APPROVED** [review]

**Pull request merged**